### PR TITLE
Fix LCAO AO MO mismatch due to particle reordering.

### LINKED ISSUE #3820: LCAO breaks when ions reordering happens
**Describe the bug**
Bug since #3809
When reordering happens, MO coefficients and AO mismatch.
See https://cdash.qmcpack.org/CDash/viewTest.php?onlyfailed&buildid=289417
short-FeCO6 and short-H2O_dimer
I will implement a fix to the AO.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Fixes #3820

## What type(s) of changes does this code introduce?
- Bugfix

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

File: src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.cpp [hunk 1]
```diff
@@ -65,13 +65,31 @@ void SoaLocalizedBasisSet<COT, ORBT>::setBasisSetSize(int nbs)
   if (BasisSetSize > 0 && nbs == BasisSetSize)
     return;
 
-  //evaluate the total basis dimension and offset for each center
-  BasisOffset[0] = 0;
-  for (int c = 0; c < NumCenters; c++)
+  if (auto& mapping = ions_.get_map_storage_to_input(); mapping.empty())
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.cpp:68]: This is intentional to confining the "mapping" inside the scope of if-statement. I would actually recommend this pattern.
I used C++17 If statement with initializer.

[the remark this answers is not in the document]
Comment (markdewing) [reply src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.cpp:68]: Ah, that makes sense. I was unaware of this c++17 feature.

**Review by prckent — APPROVED** [review]
Confirmed working:
```
Test project git_QMCPACK_prckent/build_gccnew_mkl
      Start 527: short-FeCO6_gms-vmc_noj-1-16
 1/10 Test #527: short-FeCO6_gms-vmc_noj-1-16 ...............   Passed   18.98 sec
      Start 537: short-FeCO6_pyscf-vmc_noj-1-16
 2/10 Test #537: short-FeCO6_pyscf-vmc_noj-1-16 .............   Passed   18.63 sec
      Start 528: short-FeCO6_gms-vmc_noj-1-16-kinetic
      Start 529: short-FeCO6_gms-vmc_noj-1-16-totenergy
      Start 530: short-FeCO6_gms-vmc_noj-1-16-eeenergy
      Start 531: short-FeCO6_gms-vmc_noj-1-16-samples
      Start 538: short-FeCO6_pyscf-vmc_noj-1-16-kinetic
      Start 539: short-FeCO6_pyscf-vmc_noj-1-16-totenergy
      Start 540: short-FeCO6_pyscf-vmc_noj-1-16-eeenergy
      Start 541: short-FeCO6_pyscf-vmc_noj-1-16-samples
 3/10 Test #528: short-FeCO6_gms-vmc_noj-1-16-kinetic .......   Passed    0.06 sec
 4/10 Test #529: short-FeCO6_gms-vmc_noj-1-16-totenergy .....   Passed    0.06 sec
 5/10 Test #530: short-FeCO6_gms-vmc_noj-1-16-eeenergy ......   Passed    0.06 sec
 6/10 Test #531: short-FeCO6_gms-vmc_noj-1-16-samples .......   Passed    0.06 sec
 7/10 Test #538: short-FeCO6_pyscf-vmc_noj-1-16-kinetic .....   Passed    0.06 sec
 8/10 Test #539: short-FeCO6_pyscf-vmc_noj-1-16-totenergy ...   Passed    0.06 sec
 9/10 Test #540: short-FeCO6_pyscf-vmc_noj-1-16-eeenergy ....   Passed    0.06 sec
10/10 Test #541: short-FeCO6_pyscf-vmc_noj-1-16-samples .....   Passed    0.06 sec
```

**Pull request merged**